syntax = "proto3";

package chat4all.auth.v1;

// Definição do Serviço de Autenticação
service AuthService {
    // 1. Autenticação Padrão (App/Web)
    // Valida credenciais (username/password) e retorna um JWT Token
    rpc Login (LoginRequest) returns (LoginResponse);

    // Cria um novo usuário interno (com senha)
    rpc Register (RegisterRequest) returns (RegisterResponse);

    // 2. Gestão Interna
    // Obtém dados de um usuário pelo ID (usado por outros serviços para validar existência)
    rpc GetUser (GetUserRequest) returns (User);

    // Lista usuários (Idealmente migrado para Metadata/CQRS, mas mantido aqui para Admin/Legacy)
    rpc ListUsers (ListUsersRequest) returns (ListUsersResponse);

    // 3. Integração com Connectors (WhatsApp, Telegram, etc)
    // Busca um usuário pelo ID externo (ex: telefone) ou cria um "Shadow User" automaticamente
    rpc GetOrCreateExternalUser (ExternalUserRequest) returns (ExternalUserResponse);
}

// --- MENSAGENS DE DADOS ---

message User {
    string id = 1;        // UUID interno
    string username = 2;  // username único (ou "whatsapp:+55...")
    string email = 3;
    string name = 4;
}

// --- LOGIN ---

message LoginRequest {
    string username = 1;
    string password = 2;
}

message LoginResponse {
    string access_token = 1; // JWT assinado
    string token_type = 2;   // geralmente "bearer"
}

// --- REGISTER ---

message RegisterRequest {
    string username = 1;
    string email = 2;
    string password = 3;
    string name = 4;
}

message RegisterResponse {
    User user = 1;
}

// --- GET / LIST ---

message GetUserRequest {
    string user_id = 1;
}

message ListUsersRequest {
    int32 limit = 1;
    int32 offset = 2;
}

message ListUsersResponse {
    repeated User users = 1;
}

// --- EXTERNAL USER (CONNECTORS) ---

message ExternalUserRequest {
    string provider = 1;    // ex: "whatsapp", "instagram"
    string provider_id = 2; // ex: "+5511999999999", "@usuario"
    string name = 3;        // ex: "João Silva (Via WhatsApp)"
}

message ExternalUserResponse {
    User user = 1;         // O usuário interno mapeado
    bool created_new = 2;  // True se o usuário foi criado neste momento
}